commands/src/api/camera.rs

```diff
@@ -1,12 +1,57 @@
 pub use lumeo_events::camera::Camera;
 use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use url::Url;
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub enum Request {
     /// Discover cameras. Returned data can be incomplete if camera
     /// requires authorization. For authorized access use `GetWithAuth`.
     Discover,
+
+    /// Get camera information
+    GetWithAuth {
+        url: Url,
+        credentials: Option<Credentials>,
+    },
 }
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct DiscoverResponse(pub Vec<Camera>);
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GetWithAuthResponse(pub Result<Camera, GetWithAuthError>);
+
+#[derive(Serialize, Deserialize, Debug, Clone, Error)]
+pub enum GetWithAuthError {
+    #[error("Camera is offline")]
+    CameraOffline,
+
+    #[error("Invalid credentials")]
+    InvalidCredentials,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Credentials {
+    /// Camera user name
+    pub username: String,
+
+    /// Plaintext password
+    pub password: String,
+}
```

Comment:
For pub API, could we please follow data-encapsulation conventions and have the fields private to structs (unless there is a reason not to)?

Reply:
Sorry, I don't see why we need encapsulation here.

Reply:
The same reason it's needed everywhere. If you look at API docs of `std` crate and all the most popular crates, you'll find that struct fields are typically private, unless there is a reason to not keep them private. You provide getter/setter for the fields (which could be all of them) that user could need access to.

Reply:
I think that's because std types have some behaviour. But here we have only the data.

Let's have a look at the most used crate `syn`: https://crates.io/crates?sort=recent-downloads
By coincidence this crate is all about the data too: it is about representing Rust syntax through Rust structs and enums. And all the structs in this crate have pub fields. It would be tons of boilerplate if author decided to use accessors everywhere. 

I just think we have the same situation, our crate is a set of structs that are used to be (de)serialized from JSONs coming from user side so it's all about the data, not behaviour.

Reply:
Those are very good arguments but:

* As you yourself noticed, `syn` has a huge number of structs and hiding the struct fields would have meant a lot of boilerplate indeed so I can see why they chose to just keep all fields public. As I wrote, if you've a good reason not to encapsulate, you don't and I'd call this a good reason. However we don't (and won't) have anywhere near that many structs in the crate.

* You go a bit down and you'll find a more usual crate (that's used by non-macro code), e.g: `base64`. It only has one struct that's data-only but still the [fields are kept private](https://docs.rs/base64/0.12.3/src/base64/lib.rs.html#153-160). A bit lower and you've another very popular crate, `log` that does [the same](https://docs.rs/log/0.4.11/src/log/lib.rs.html#1039-1042). I didn't go further but I'm sure you'll find the same in most crates.

As I explained to @jplatte in another PR, one of the main benefits of data encapsulation is that the exact field types remain hidden details that you can change w/o breaking the public API. E.g here, the getters will return an `&str` so in future if we switch to a different container type for the string, the API won't break. I know this is not a great example but suppose we wanted to make this crate work on a low-end edge device and we had to switch to using [ArrayString](https://docs.rs/arrayvec/0.5.1/arrayvec/struct.ArrayString.html) type.

Reply:
Well, on that other PR, I also didn't agree. I just wasn't feeling like discussing it more. However, now that it's coming up again, I guess we might need to discuss this in detail. I understand where you are coming from @zeenix; in the grand scheme of things having public fields in object-oriented or similar languages is a very uncommon thing.

I tried this in C++ some time ago and had to realize it just doesn't work nicely with some parts of the language¹, in most other object-oriented languages, everything is mutable by default and usually having a const / immutable / final `Foo` doesn't actually mean that you can't mutate its fields.

However, I have yet to see these same issues in Rust. In [Ruma](https://github.com/ruma), we also have API definition structs (for a REST API, contrary to the `syn` example) with public fields and it hasn't resulted in any issues so far. I could see this applying to fields with internal mutability, and thinking about it I don't think I would ever make such fields public in library code. But that's not what we're dealing with here.

How about this: We're still rather early in our development. Breaking change releases won't hurt so much. We continue with public fields for now, and if it does end up causing issues, we can re-evaluate. Maybe at that point there's more tooling to derive accessors, or support in rust-analyzer to generate them, and it's going to be one easy refactoring session.

¹ in particular move constructors

Reply:
> Well, on that other PR, I also didn't agree. I just wasn't feeling like discussing it more.

Yeah I know but as they say "silence is compliance" so if say you don't care enough to continue discussion, the person who care more gets to make the decision. :)

> I have yet to see these same issues in Rust.

I pointed out to at least 2 very popular crates that hide their struct internals even when there is no behaviour associated with them. So I really don't buy this "this is not a problem in Rust" argument.

> In Ruma, we also have API definition structs (for a REST API, contrary to the syn example) with public fields and it hasn't resulted in any issues so far.

Well the issues I mentioned are there for sure, just that they haven't materialized. It might never materialize for us either but if they do, it'll mean breaking the API. I gave some examples in my previous comment how that can happen.

> How about this: We're still rather early in our development. Breaking change releases won't hurt so much. We continue with public fields for now, and if it does end up causing issues, we can re-evaluate. Maybe at that point there's more tooling to derive accessors, or support in rust-analyzer to generate them, and it's going to be one easy refactoring session.

The problem is that the issues associated with exposing struct internals, are mainly going to realize when/if other people start using them and most likely by then our API will have to be stable (enough). We can always change the API and update our using code at the same time so won't be an issue for us.

Having said all that, if you folks are insistent on this, I'm not going to argue anymore. Just that when/if any issues arise from this, I will be there to tell you "I TOLD YOU SO!". :)

Reply:
@zeenix I almost agreed :)

Well, looks like everyone knows what is encapsulation about and what are tradeoffs. The only question is how do we weigh them. 

This of course can be done but for sure in another PR and for all structs, not only these ^^^, which are not user-facing btw and are used only for MQTT comms. It is relatively easy to make a proc-macro to generate getters/setters but we cannot generate docs (not sure). Also we need to derive a Builder.

Too much to think about for this tiny PR though.

Reply:
> This of course can be done but for sure in another PR and for all structs, not only these

👍 

> It is relatively easy to make a proc-macro to generate getters/setters but we cannot generate docs (not sure).

It's actually not at all a problem to generate docs from a proc-macro, in the form of `#[doc]` attributes, which is equivalent to `///` comments. Also, doc comments present on types that use a derive macro will be transformed into doc attributes first so it's also straight-forward to inspect the docs of a type or its fields in a proc macro.